protobuf_serialization/reader.nim

```diff
@@ -42,7 +43,54 @@ proc readFieldInto[T: object](
       raise (ref ValueError)(msg: "not enough bytes")
     memoryInput(tmp).readValueInternal(value)
 
-proc readFieldInto[T: not object and (seq[byte] or not seq)](
+when defined(ConformanceTest):
+  proc readFieldInto[K, V](
+    stream: InputStream,
+    value: var Table[K, V],
+    header: FieldHeader,
+    ProtoType: type
+  ) =
+    # I know it's ugly, but I cannot find a clean way to do it
+    # ... And nobody cares about map
+    when K is SomePBInt and V is SomePBInt:
+      type
+        TableObject {.proto3.} = object
+          key {.fieldNumber: 1, pint.}: K
+          value {.fieldNumber: 2, pint.}: V
+    elif K is SomePBInt:
+      type
+        TableObject {.proto3.} = object
+          key {.fieldNumber: 1, pint.}: K
+          value {.fieldNumber: 2.}: V
+    elif V is SomePBInt:
+      type
+        TableObject {.proto3.} = object
+          key {.fieldNumber: 1.}: K
+          value {.fieldNumber: 2, pint.}: V
+    else:
+      type
+        TableObject {.proto3.} = object
+          key {.fieldNumber: 1.}: K
+          value {.fieldNumber: 2.}: V
+    var tmp = default(TableObject)
+    stream.readFieldInto(tmp, header, ProtoType)
+    value[tmp.key] = tmp.value
+
+proc readFieldInto[T: enum](
+  stream: InputStream,
+  value: var T,
+  header: FieldHeader,
+  ProtoType: type
+) =
+  # TODO: This function doesn't work for proto2 edge cases. Make it work
+  when 0 notin T and T.isProto3():
+    {.fatal: $T & " definition must contain a constant that maps to zero".}
+  header.requireKind(WireKind.Varint)
+  let enumValue = stream.readValue(ProtoType)
+  if not checkedEnumAssign(value, enumValue.int32):
+    discard checkedEnumAssign(value, 0)
```

Comment:
Herm, actually, is this compatible with proto2?
I can't find info on how to handle invalid values & their interaction with optional for instance

Reply:
similar to what we do for other specs, and a few places in this file, it would be good to include a spec link for every significant decision or function (like so: https://github.com/status-im/nim-protobuf-serialization/blob/bd4520d96d0017df67490eeec6d3912b2d3f23a4/protobuf_serialization/codec.nim#L37) - enum in particular did change behavior between 2 & 3 which makes it more difficult to support / translate to `nim` enum

Reply:
Seems like in proto2, the default is the first value (which may not be zero)

Reply:
So:
proto2:
No specific requirements on the enum
Defaults to first value*
Unknown value treated like unknown field (So for Optional, should be none. For required, should fail parsing)*

proto3:
First element must be 0*
Defaults to first value / 0
Unknown value must be mapped to a specific value in the enum (ie, the proto3 macro should add a UNKNOWN field to every enum)*

`*` = things we are missing AFAICT